Added Docker image for building from source

gracewgao opened the pull request:
Here's the Dockerfile and the associated documentation for building pyre-check from source within a Docker container. This setup has already been tested on my Apple M1, but let me know if there's anything else I should change/add!

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+ENV PYRE_BINARY='/home/opam/pyre-check/source/_build/default/main.exe'
+ENV PYRE_TYPESHED='/home/opam/pyre-check/stubs/typeshed/typeshed-master'
+ENV PYTHONPATH='/home/opam/pyre-check:$PYTHONPATH'
+ENV PATH='/home/opam/.opam/4.10.2/bin:/home/opam/pyre-check/scripts:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH'
```
Wow, are `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin` all really missing from the `PATH` by default?

gracewgao replied:
I'm not sure but it seemed unable to find programs like make etc. unless I did this? I think it's maybe because the $PATH variable isn't evaluating properly

gbleaney replied:
@gracewgao I think the problem is that you're using single quotes, not double quotes. You need double quotes for string interpolation to occur.

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+ENV PYTHONPATH='/home/opam/pyre-check:$PYTHONPATH'
+ENV PATH='/home/opam/.opam/4.10.2/bin:/home/opam/pyre-check/scripts:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH'
+
+RUN echo -e 'export PYRE_BINARY=/home/opam/pyre-check/source/_build/default/main.exe \n \
```
Why do we need to both set these using ENV and put them in `.bashrc`? Shouldn't it be one or the other? If for some reason there is a process which doesn't inherit the variables through `ENV` but does pick them up from `.bashrc`, couldn't we just eliminate the duplicate `ENV` commands by moving to `RUN source ~/.bashrc`?

gracewgao replied:
I tried that first but it wasn't working, so I found some way online to define the environment variables in Docker using ENV - I added them into .bashrc afterwards because I figured they would need to be reevaluated every time they restart the container?

gbleaney replied:
It looks like variables set using `ENV` are supposed to persist when the container is run:
https://docs.docker.com/engine/reference/builder/#env

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+SHELL ["/bin/bash", "-c"]
+
+# updates environment variables
+ENV PYRE_BINARY='/home/opam/pyre-check/source/_build/default/main.exe'
```
Nit: Let's define these at the end when these executables actually exist

gracewgao replied:
Yep, I'll change that!

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+# installs system dependencies
+RUN sudo apt update && sudo apt upgrade -y \
+    && sudo apt install python3 python3-pip pkg-config -y \
+    && sudo apt-get install libsqlite3-dev build-essential -y \ 
```
Why `apt` in some cases and `apt-get` in others?

gracewgao replied:
Honestly not too sure - which one would you recommend using?

gbleaney replied:
IIRC `apt` is the newer, simpler, more user friendly wrapper. I don't have a strong preference either way, but I would prefer we be consistent

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+
+# downloads and installs pyre-check
+RUN git clone https://github.com/facebook/pyre-check \
+    && cd pyre-check \
```
Prefer `WORKDIR` over `cd`:
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#workdir

(Apply this throughout)

gracewgao replied:
Thanks, I'll change that!

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    && opam switch create 4.10.2
+
+# downloads and installs pyre-check
+RUN git clone https://github.com/facebook/pyre-check \
```
Given that the current user will have to have this repo checked out to get this dockerfile, we should use the working context rather than pulling in from the web (this has the additional benefit of working with local changes):
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#understand-build-context

You will probably want to add a `.dockerignore` as well, so avoid pulling in build artifacts and the like:
https://docs.docker.com/engine/reference/builder/#dockerignore-file

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    export PATH=/home/opam/pyre-check/scripts:$PATH' >> ~/.bashrc
+
+# installs system dependencies
+RUN sudo apt update && sudo apt upgrade -y \
```
Commands like this that take a long time and don't change often should be as early in the file as possible to take advantage of build caching:
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    && make test
+
+# creates scripts for python alias and pyre-check
+RUN sudo touch /usr/bin/python \
```
I think these four lines be done by creating a symlink? Eg. my `python` command currently is a symlink:
```
(venv) gbleaney-mbp:python_security gbleaney$ which python
/Users/gbleaney/python_security/venv/bin/python
(venv) gbleaney-mbp:python_security gbleaney$ ll /Users/gbleaney/python_security/venv/bin/python
lrwxr-xr-x  1 gbleaney  1876110778  7 Mar 31 23:43 /Users/gbleaney/python_security/venv/bin/python -> python3
```

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+
+# runs python tests
+RUN cd /home/opam/pyre-check \ 
+    && pip3 install -r requirements.txt \
```
Same comment as on the `apt` step. Installing dependencies should probably go earlier

gracewgao replied:
We can't install the dependencies until the requirements.txt inside the repo is cloned, so it would still have to be below that? I put them where they are now because the pip dependencies aren't really used for anything other than the Python tests, but I can change that if you want.

gbleaney replied:
Generally you want expensive, infrequently changed steps higher in the file. So I'd say the correct order would be
1. apt installs
2. bring in requirements.txt
3. pip install
4. bring in typeshed and unzip
4. bring in rest of repo
5. build

Doing it that way means you never have to re-run the install or unzip steps, even if you're actively editing and changing code in the repo and re-launching the container

gracewgao replied:
Okay, that makes sense - I'll change that now!

gbleaney commented on CONTRIBUTING.md:
```diff
@@ -85,6 +85,39 @@ During analysis, each function will be processed into a control flow graph to re
 
+### Docker Setup
+
+If you're having issues setting up or your OS is not yet supported, you can also use a Docker image. It runs [Debian GNU/Linux 10 (buster)](https://www.debian.org/) and builds pyre-check from source. 
```
Let's not duplicate docs like this. Instead, let's put a pointer here to the other docs. That helps users find them, without adding extra maintenance overhead

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,51 @@
+RUN cd /home/opam/pyre-check/stubs/typeshed \
+    && unzip typeshed.zip
+
+CMD /bin/bash
```
Is this actually needed? I see in the instructions below, you have people also specify `/bin/bash` as the shell to use

gracewgao replied:
Hmm, I just got rid of the other instances where I specified /bin/bash and just kept this one

gbleaney commented on documentation/website/docs/installation.md:
```diff
@@ -86,6 +86,40 @@ $ echo 'PATH="/path/to/pyre-check/scripts:$PATH"' >> ~/.bashrc
+
+4. Inside the container, run any pyre-check command now with `pyre`!
+
+   *Note: When initializing pyre-check with `pyre init`, you may have to enter the following paths for the binary and typeshed:*
```
Shouldn't these have been set in the dockerfile? Are these still required here?

gracewgao replied:
I think you have to set it up in the Dockerfile but then it asks for it again when initializing pyre?

gbleaney replied:
Ugh, we should fix that on the Pyre side, but that's out of scope here

gbleaney reviewed (changes requested):
Thanks for the PR! I'm really excited to have this landed as an alternative way to build pyre

gracewgao commented:
I've made some changes that you suggested - let me know if there's anything else?

gbleaney commented on Dockerfile:
```diff
@@ -0,0 +1,45 @@
+ENV PATH="/home/opam/.opam/4.10.2/bin:$PATH"
+
+# clones pyre-check and installs dependencies
+RUN git clone https://github.com/facebook/pyre-check
```
We should use `ADD` to grab files from the local filesystem:
https://docs.docker.com/engine/reference/builder/#add

Note that we probably want to do this in three steps, to prevent misc file changes from causing the unzip and pip install steps from running:
1. Pull in zip file (then unzip)
2. Pull in requirements.txt (then pip install)
3. Pull in remaining files

gbleaney reviewed (changes requested):
Getting really close! Mostly just docs tweaks at this point

gracewgao commented:
Just updated the changes you suggested! The ADD command took a bit of permission toggling but the current setup works for me. Regarding this, I also added a quick note in the installation docs that the pyre-check files should only be changed locally and then rebuilt in the Docker image.

gbleaney commented:
@gracewgao I'm putting this through our internal review system now. In the mean time, can you please add some evidence of testing?

gracewgao commented:
@gbleaney Sure! Apart from the integration tests and tests run inside the Docker container, would you like me to manually test it and attach screenshots?

arthaud commented on documentation/website/docs/installation.md:
````diff
@@ -86,6 +86,43 @@ $ echo 'PATH="/path/to/pyre-check/scripts:$PATH"' >> ~/.bashrc
+
+3. Run the new image in a new container `pyre-container` (or another name if you wish)
+   ```bash
+   docker run \                           
````
Here and other places: please remove those trailing whitespaces
```suggestion
   docker run \
```

arthaud commented on documentation/website/docs/installation.md:
````diff
@@ -86,6 +86,43 @@ $ echo 'PATH="/path/to/pyre-check/scripts:$PATH"' >> ~/.bashrc
+   ƛ Unable to locate typeshed, please enter its root: /home/opam/pyre-check/stubs/typeshed/typeshed-master
+   ```
+
+**For contributors:** Inside the Docker container, the added pyre-check directory is only editable by the root user. To contribute to Pyre, make edits in your local filesystem and rebuild the Docker by running Step 2, then running a new DOcker container in Steps 3-4.
````
Typo
```suggestion
**For contributors:** Inside the Docker container, the added pyre-check directory is only editable by the root user. To contribute to Pyre, make edits in your local filesystem and rebuild the Docker by running Step 2, then running a new Docker container in Steps 3-4.
```

gracewgao commented:
@arthaud Thanks for catching that - I just fixed up the whitespace and typo!
